flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/ParquetTableSource.java

```diff
@@ -0,0 +1,539 @@
+	}
+
+	private ParquetTableSource(String path, MessageType parquetSchema, Configuration configuration,
+									boolean recursiveEnumeration, int[] selectedFields, FilterPredicate predicate) {
+		Preconditions.checkNotNull(path, "Path must not be null.");
+		Preconditions.checkNotNull(parquetSchema, "ParquetSchema must not be null.");
+		Preconditions.checkNotNull(configuration, "Configuration must not be null");
+		this.path = path;
+		this.parquetSchema = parquetSchema;
+		this.parquetConfig = configuration;
+		this.selectedFields = selectedFields;
+		this.predicate = predicate;
+		this.recursiveEnumeration = recursiveEnumeration;
+
+		// determine the type information from the Parquet schema
+		RowTypeInfo typeInfoFromSchema = (RowTypeInfo) ParquetSchemaConverter.fromParquetType(parquetSchema);
+
+		// set return type info
+		if (selectedFields == null) {
+			this.typeInfo = typeInfoFromSchema;
+		} else {
+			this.typeInfo = RowTypeInfo.projectFields(typeInfoFromSchema, selectedFields);
+		}
+
+		// create a TableSchema that corresponds to the Parquet schema
+		this.tableSchema = new TableSchema(
+			typeInfoFromSchema.getFieldNames(),
+			typeInfoFromSchema.getFieldTypes()
+		);
+	}
+
+	@Override
+	public TableSource<Row> projectFields(int[] fields) {
+		return new ParquetTableSource(path, parquetSchema, parquetConfig, recursiveEnumeration, fields, null);
+	}
+
+	@Override
+	public DataSet<Row> getDataSet(ExecutionEnvironment executionEnvironment) {
+		ParquetRowInputFormat parquetRowInputFormat = buildParquetInputFormat();
+		parquetRowInputFormat.setNestedFileEnumeration(recursiveEnumeration);
+		if (selectedFields != null) {
+			parquetRowInputFormat.selectFields(typeInfo.getFieldNames());
+		}
+
+		if (predicate != null) {
+			parquetRowInputFormat.setFilterPredicate(predicate);
+		}
+		return executionEnvironment.createInput(parquetRowInputFormat).name(explainSource());
+	}
+
+	@Override
+	public TableSource<Row> applyPredicate(List<Expression> predicates) {
+		FilterPredicate predicate = null;
+		if (predicates.size() > 0) {
+			predicate = toParquetPredicate(predicates.get(0));
+
+			for (Expression expression : predicates.subList(1, predicates.size())) {
+				FilterPredicate right = toParquetPredicate(expression);
+				if (predicate != null && right != null) {
+					predicate = FilterApi.and(predicate, right);
+				} else if (right != null) {
+					predicate = right;
+				}
+			}
+		}
+
+		return new ParquetTableSource(path, parquetSchema, this.parquetConfig, recursiveEnumeration, selectedFields, predicate);
+	}
+
+	@Override
+	public boolean isFilterPushedDown() {
+		return predicate != null;
+	}
+
+	@Override
+	public TypeInformation<Row> getReturnType() {
+		return typeInfo;
+	}
+
+	@Override
+	public TableSchema getTableSchema() {
+		return tableSchema;
+	}
+
+	@Override
+	public String explainSource() {
+		return "ParquetFile[path=" + path + ", schema=" + parquetSchema + ", filter=" + predicateString() + "]";
```

Comment:
Include the projected fields in the source description. Flink checks whether the project pushdown worked by comparing the descriptions. If both are the same, the rule is not applied.